### PR TITLE
copr: Add vectorize GreatestInt and LeastInt

### PR DESCRIPTION
**Pull request by Aloxaf**
Signed-off-by: Aloxaf <aloxafx@gmail.com>

PCP #5751

###  What have you changed?

Add vectorize `greatest_int` and `least_int`

###  What is the type of the changes?

- Improvement (a change which is an improvement to an existing feature)

###  How is the PR tested?

- Unit test
- Integration test

###  Does this PR affect documentation (docs) or should it be mentioned in the release notes?

No

###  Does this PR affect `tidb-ansible`?

No

### CODE REVIEW

**Comment by Aloxaf** [comment]
@sre-bot /run-integration-copr-test copr-test=pr/106

File: components/tidb_query/src/rpn_expr/impl_compare.rs [hunk 1]
```diff
@@ -207,10 +207,46 @@ pub fn coalesce<T: Evaluable>(args: &[&Option<T>]) -> Result<Option<T>> {
     Ok(None)
 }
 
+#[inline]
+fn do_get_extremum<F, T>(list: &[&Option<T>], chooser: F) -> Result<Option<T>>
```
Comment (breeswish) [inline components/tidb_query/src/rpn_expr/impl_compare.rs]: Looks like we can extract metadata for greatest and least? @TennyZhuang

Comment (TennyZhuang) [reply components/tidb_query/src/rpn_expr/impl_compare.rs]: Yes, we can extract greatest and least in constants and save it in metadata
.

Comment (breeswish) [reply components/tidb_query/src/rpn_expr/impl_compare.rs]: Hi @Aloxaf , thanks for the contribution! According to @TennyZhuang this PR can be optimized by extracting the extremum constant value beforehand. However currently we are refining related part. You may wait https://github.com/pingcap/tidb/pull/13962 being merged and use the newer solution later.

Comment (Aloxaf) [reply components/tidb_query/src/rpn_expr/impl_compare.rs]: @breeswish Thanks for your advice! I'll wait for it.

Comment (breeswish) [reply components/tidb_query/src/rpn_expr/impl_compare.rs]: Feel free to pick up some other easy functions if you want :)

**Comment by Aloxaf** [comment]
Ummm, I'm still a little confused about what I should do.

Should I write something like
```rust
#[rpn_fn(varg, capture = [metadata], min_args = 1, metadata_mapper = init_extremum_data::<Int>)]
#[inline]
pub fn greatest_int(metadata: &ExtremumData<Int>, args: &[&Option<Int>]) -> Result<Option<Int>> {
    Ok(Some(metadata.get_greatest()))
}
```
??

**Comment by breeswish** [comment]
@Aloxaf Sorry, @TennyZhuang is still refactoring his implicit arg PR since it is also blocked by something else. Although the framework is finished, you may still want to wait for his PR so that you will have an *example* to know how to do.

**Comment by Aloxaf** [comment]
@breeswish  OK, thanks for your reply.

**Comment by breeswish** [comment]
I guess tennyzhuang cannot finish it in short time. Maybe we should proceed first without using the optimization? @lonng what do you think

**Comment by lonng** [comment]
@breeswish @Aloxaf We can complete this task first and leave the optimization in the future.

**Comment by andylokandy** [comment]
@breeswish @lonng PTAL

**Comment by TennyZhuang** [comment]
Please resolve the conflict

**Comment by AndreMouche** [comment]
@breeswish @lonng PTAL and the next deadline for reviewing is 2020-02-20 24:00:00

**Comment by TennyZhuang** [comment]
lint is failed

**Comment by AndreMouche** [comment]
Please fix the lint

**Comment by Aloxaf** [comment]
Sorry, didn't see it.

File: components/tidb_query/记录.txt [hunk 2]
```diff
@@ -0,0 +1 @@
+rg '^ +([a-zA-Z]+) => ([a-z_]+),' src/expr/scalar_function.rs -r '$1 $2' | % { $w=$_.split(' '); ((rg "fn $($w[1])" ./src/rpn_expr/ | Out-Null) || (rg "\[x\].*``$($w[0])`` \(Ve" /tmp/tmp.txt | Out-Null )) || echo $w[1] } | sort
```
Comment (breeswish) [inline components/tidb_query/记录.txt]: Please remove this file

File: components/tidb_query/src/rpn_expr/impl_compare.rs [hunk 3]
```diff
@@ -849,4 +885,51 @@ mod tests {
...
+            (vec![Some(0), Some(4), Some(8)], Some(8), Some(0)),
+        ];
+
+        type Cases<T> = Vec<(Vec<Option<T>>, Option<T>, Option<T>)>;
+
+        fn do_test<T>(greatest_sig: ScalarFuncSig, least_sig: ScalarFuncSig, cases: Cases<T>)
+        where
+            T: Debug + PartialEq + Evaluable,
+            Option<T>: From<ScalarValue> + Into<ScalarValue>,
+        {
```
Comment (breeswish) [inline components/tidb_query/src/rpn_expr/impl_compare.rs]: you can simply remove these lines and write `ScalarFuncSig::GreatestInt` inline directly.

File: components/tidb_query/记录.txt (hunk 2 above)
Comment (Aloxaf) [reply components/tidb_query/记录.txt]: Oh sorry, added by mistake.

File: components/tidb_query/src/rpn_expr/impl_compare.rs (hunk 3 above)
Comment (Aloxaf) [reply components/tidb_query/src/rpn_expr/impl_compare.rs]: finished

**Comment by andylokandy** [comment]
@Aloxaf Sorry for the delay, would you like to resolve the conflicts? I'd merge it soon after that.

**Comment by Aloxaf** [comment]
It seems that this part is already finished in the master branch.

I think this PR can be closed now.

**Pull request closed without merging**